File: AEDII/AED2-FELIPE-24-1/TP01/Q10/Q10.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>

bool ehPalindromoRec(char palavra[], int inicio, int fim)
{
    if (inicio >= fim)
    {
        return true;
    }
    if (palavra[inicio] != palavra[fim])
    {
        return false;
    }
    return ehPalindromoRec(palavra, inicio + 1, fim - 1);
}

bool ehPalindromo(char palavra[])
{
    int tamanho = strlen(palavra);
    return ehPalindromoRec(palavra, 0, tamanho - 1);
}

int main()
{

    char *palavra = (char *)malloc(200 * sizeof(char));
    bool palindromo = false;
    scanf(" %[^\r\n%*c]", palavra);
    while (strcmp(palavra, "FIM") != 0)
    {
        palindromo = ehPalindromo(palavra);
        if (palindromo)
        {
            printf("SIM\n");
        }
        else
        {
            printf("NAO\n");
        }
        scanf(" %[^\r\n]%*c", palavra);
    }
    return 0;
}

File: 2024/Estruturas de Dados e Algoritmos I-II/AEDII/AED2-FELIPE-24-1/TP01/Q10/Q10.java
import java.io.*;
public class Q10 {

    public static void main(String[] args) {

        String palavra = MyIO.readLine();
        while (!palavra.equalsIgnoreCase("FIM")) {
            if (isPalindrome(palavra, 0, palavra.length() - 1)) {
                System.out.println("SIM");
            } else {
                System.out.println("NAO");
            }
            palavra = MyIO.readLine();
        }

    }

    public static boolean isPalindrome(String palavra, int inicio, int fim) {
        if (inicio >= fim) {
            return true;
        }
        if (palavra.charAt(inicio) != palavra.charAt(fim)) {
            return false;
        }
        return isPalindrome(palavra, inicio + 1, fim - 1);
    }
}
